core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/BillingAccountChangeSynchronizer.scala

```diff
@@ -0,0 +1,312 @@
+import org.broadinstitute.dsde.rawls.model._
+import org.broadinstitute.dsde.rawls.monitor.migration.MigrationUtils.Implicits._
+import org.broadinstitute.dsde.rawls.monitor.migration.MigrationUtils.Outcome
+import org.broadinstitute.dsde.rawls.monitor.migration.MigrationUtils.Outcome.{Failure, Success}
+
+import java.time.Instant
+import scala.concurrent.duration._
+
+object BillingAccountChangeSynchronizer {
+  sealed trait WorkspaceBillingAccountsMessage
+
+  case object UpdateBillingAccounts extends WorkspaceBillingAccountsMessage
+
+  def apply(dataSource: SlickDataSource,
+            gcsDAO: GoogleServicesDAO,
+            samDAO: SamDAO,
+            initialDelay: FiniteDuration,
+            pollInterval: FiniteDuration): Behavior[WorkspaceBillingAccountsMessage] =
+    Behaviors.setup { context =>
+      val actor = BillingAccountChangeSynchronizer(dataSource, gcsDAO, samDAO)
+      Behaviors.withTimers { scheduler =>
+        scheduler.startTimerAtFixedRate(UpdateBillingAccounts, initialDelay, pollInterval)
+        Behaviors.receiveMessage {
+          case UpdateBillingAccounts =>
+            try actor.updateBillingAccounts.unsafeRunSync()
+            catch {
+              case t: Throwable => context.executionContext.reportFailure(t)
+            }
+            Behaviors.same
+        }
+      }
+    }
+}
+
+final case class BillingAccountChangeSynchronizer(dataSource: SlickDataSource,
+                                                  gcsDAO: GoogleServicesDAO,
+                                                  samDAO: SamDAO
+                                                 )
+  extends LazyLogging {
+
+  import dataSource.dataAccess._
+  import dataSource.dataAccess.driver.api._
+
+  /* Sync billing account changes to billing projects and their associated workspaces with Google
+   * one at a time.
+   *
+   * Why not all at once?
+   * To reduce the number of stale billing account changes we make.
+   *
+   * Let's imagine that there are lots of billing account changes to sync. If a user changes
+   * the billing account on a billing project that's currently being sync'ed then we'll have to
+   * sync the billing projects and all workspaces again immediately. If we sync one at a time
+   * we'll have a better chance of doing this once.
+   */
+  def updateBillingAccounts: IO[Unit] =
+    readABillingProjectChange.flatMap(_.traverse_ {
+      syncBillingAccountChange[Kleisli[IO, BillingAccountChange, *]]
+        .handleErrorWith(failChange[Kleisli[IO, BillingAccountChange, *]])
+        .run
+    })
+
+
+  def readABillingProjectChange: IO[Option[BillingAccountChange]] =
+    inTransaction {
+      BillingAccountChanges
+        .latestChanges
+        .unsynced
+        .take(1)
+        .result
+    }.map(_.headOption)
+
+
+  private def syncBillingAccountChange[F[_]](implicit R: Ask[F, BillingAccountChange], M: MonadThrow[F], L: LiftIO[F])
```

Comment:
Good observation -  the `BillingAccountChange` is effectively passed around implictly for the only reason that it removes one argument everywhere. This allows me to log information without being explict about all the information I'm logging (see the logContext funciton below).
`Ask` is a generalisation of `Reader` that you met in the `WorkspaceMigrationActor`. It abstracts dependecy injection such that I can pass in a `Reader[BillingAccountChange]` OR i can use a function that suppplies a `BilllingAccountChange` or I supply any structure that allows me to fetch a billing account change.
It's pretty neat and while I haven't made the most of what `Ask` has to offer, it allows me to abstract the concrete type I flatmap over. I wish I had used this in the migration code tbh.
That's rather a lot - I'd be glad to discuss it futher in a call :)

Reply:
I thought I'd have a second attempt at answering this because you've asked one of those simple questions that has no simple answer. I'll try to be brief but the scope of this question spans many topics including what cats is trying to bring to scala - ie type classes. I know I'm going to make a mess of this...

The `BillingAccountChange` is not implict. What this is saying that the implict parameter `R` supplies the `Ask[_, BillingAccountChange]` type class instance for the type `F`. I've described `Ask` above but the cat's documentation does a good job at explaining it I think https://typelevel.org/cats-mtl/mtl-classes/ask.html.

So... What's a type class?
I'll give a hand-wavey description of type classes (and how they differ from object-oriented classes). I thoroughly recommend the following sources for this topic to understand what type classes *really* are:
https://youtu.be/re96UgMk6GQ?t=2583
https://homepages.inf.ed.ac.uk/wadler/papers/class-letter/class-letter.txt

The problem: How can we generalise over the types of object certain operations apply to? 

Consider sorting a list of objects. How can we say "this function is defined for all orderable objects"? How do we encode a generic "compare" operation?
An object-oriented approach might be to define the interface `Ordered` and to have all orderable types implement that interface. In Scala, we might write
```scala
// interface
trait Ordered[A] {
  def compare(another: A): Int
}

// sub-type
class MyType extends Ordered[MyType] {
  override def compare(another: MyType): Int = ???
}

def sort[A: Ordered](xs: List[A]): List[A] = ???
```
This approach is called "sub-typing" - the generic operation `compare` is defined on a type if that type is a sub-type of (extends) the interface that defines that operation. Generic operations are defined internally to the type (ie within the definition of that type) and are often used in a hierarchy of types. Which definition to call is resolved at runtime via https://en.wikipedia.org/wiki/Dynamic_dispatch.

Purely functional languages like Haskell were designed without object hierarchies. Instead, they encode the availabilty of a generic operation by defining a class of types (think "set") that obay certain laws or provide implementations for generic functions. Instead of saying that type X "is a" Y, we say type X is an "instance" or "member of" class of types Y. In Scala, we could write: 
```scala
case class MyType { .. }

// type class
trait Orderable[A] {
  def compare(a: A, b A): Int
}

// type class instance (or how it's a member of the class of types)
implict object orderableMyType extends new Orderable[MyType] {
  override def compare(a: MyType, b: MyType) = ???
}

def sort[A](xs: List[A])(implict O: Orderable[MyType]): List[A] = ???
```
There are a few important observations here, one is common to type classes and the others are specific to Scala's (or cat's) implementation of type-classes:
1. Type classes provide ad-hoc polymorphism - support for generic operations does not need to be baked into the type's definition (unlike subtyping).
2. Type classes in Scala are implemented by subtyping because Scala is an object-oriented language and uses dynamic dispatch to resolve method calls.
3. Type class support in scala is supplied by an implict parameter on which generic type-class operations are dispatched, ie 
```scala
// subtyping
def greaterThan[A: Ordered](a: A, b: A): Boolean = 
  a.compare(b) > 0

// type class
def greaterThan[A](a: A, b: A)(implict O: Orderable[A]): Boolean =
  O.compare(a, b) > 0
```

Key take aways:
Type classes provide a means of defining generic operations and ad-hoc implementations without modifying the type's definition.
Types do not implement type classes in their definition. Instead an instance of that type class for that type is defined externally.
In Scala, type class instances are often defined as `implict` and passed as implict parameters to a function.